astroplan/core.py

```diff
@@ -341,30 +426,61 @@ def _horiz_cross(self, t, alt, rise_set, horizon=0*u.degree):
+            alt = alt[np.newaxis, :]
+        n_targets = alt.shape[0]
+
         if rise_set == 'rising':
             # Find index where altitude goes from below to above horizon
-            condition = (alt[:-1] < horizon) * (alt[1:] > horizon)
+            condition = (alt[:, :-1] < horizon) * (alt[:, 1:] > horizon)
         elif rise_set == 'setting':
             # Find index where altitude goes from above to below horizon
-            condition = (alt[:-1] > horizon) * (alt[1:] < horizon)
+            condition = (alt[:, :-1] > horizon) * (alt[:, 1:] < horizon)
+
+        target_inds, time_inds = np.nonzero(condition)
+
+        if np.count_nonzero(condition) < n_targets:
+            target_inds, _ = np.nonzero(condition)
+            noncrossing_target_ind = np.setdiff1d(np.arange(n_targets),
+                                                  target_inds,
+                                                  assume_unique=True)#[0]
 
-        if not np.any(condition):
-            warnmsg = ('Target does not cross horizon={} within 24 '
-                       'hours'.format(horizon))
-            if (alt > horizon).all():
+            warnmsg = ('Target(s) index {} does not cross horizon={} within '
+                       '24 hours'.format(noncrossing_target_ind, horizon))
+
+            if (alt[noncrossing_target_ind, :] > horizon).all():
                 warnings.warn(warnmsg, TargetAlwaysUpWarning)
             else:
                 warnings.warn(warnmsg, TargetNeverUpWarning)
-            return (None, None), (None, None)
-
-        # Isolate horizon crossing
-        nearest_index = np.argwhere(condition)[0][0]
-
-        # Capture points on either side of horizon crossing
-        lower_t, upper_t = t[nearest_index:nearest_index+2]
-        lower_alt, upper_alt = alt[nearest_index:nearest_index+2]
 
-        return (lower_t, upper_t), (lower_alt, upper_alt)
+            # Fill in missing time with np.nan
+            target_inds = np.insert(target_inds, noncrossing_target_ind,
+                                    noncrossing_target_ind)
+            time_inds = np.insert(time_inds.astype(float),
+                                  noncrossing_target_ind,
+                                  np.nan)
+        elif np.count_nonzero(condition) > n_targets:
+            dup_target_inds = list(set([target_ind for target_ind in target_inds
+                                        if np.sum(target_inds == target_ind) > 1]))
+            old_target_inds = np.copy(target_inds)
+            old_time_inds = np.copy(time_inds)
+
+            time_inds = []
+            target_inds = []
+            for tgt, tm in zip(old_target_inds, old_time_inds):
+                if tgt not in target_inds:
+                    time_inds.append(tm)
+                    target_inds.append(tgt)
+            target_inds = np.array(target_inds)
+            time_inds = np.array(time_inds)
+
+        times = [t[i:i+2] if not np.isnan(i) else i for i in time_inds]
```

Comment:
Hmm, I'm not totally convinced this and the line below are right. I need some time to play around with this some more...